### ISSUE TITLE
KeyError: 'Name', BlueZ API yields invalid responses

### ISSUE THREAD

**Issue by JakkFish** [Poohl/joycontrol#6]
![755C24A1-3F49-4704-A81A-E8F957429F01](https://user-images.githubusercontent.com/84202439/118316134-0e32e580-b4bc-11eb-9cf0-ec06a168d676.jpeg)

**Comment by Poohl** [comment]
@JakkFish weird, will look into it. This is some convinience code to sanity check the given parameters.
The hotfix: Go into `joycontrol/joycontrol/device.py` and change the `get_paired_switches(self)` definition to be the following:
```python
    def get_paired_switches(self):
        return []
```
This will cause the pairing-detection to fail, meaning `-r auto` and unpairing sanity-check will no longer work.

A few more questions:
- OS
- BT-controller
- bluez-version
- If you want to, try using `d-feet` (might have to install) to look for some object with the `org.bluez.Device1` interface inside the `org.bluez` service on the System-bus that has no `Name` Property.

**Comment by JakkFish** [comment]
Okay i will try the code in a second also my OS is Ubuntu 20.04.2 LTS and im using Cambridge Silicon Radio, Ltd Bluetooth Dongle (HCI mode) [8891] My bluez version 5.53 and i searched d-feet and im not sure what you mean but i have a picture im going to send of what i see

**Comment by JakkFish** [comment]
![A8459A26-CD86-4913-A2EA-54B6FA0951B3](https://user-images.githubusercontent.com/84202439/118332974-873c3800-b4d0-11eb-86bb-5551bfe386e9.jpeg)
![0277C84F-82D5-4BB6-9172-32FF2F7D5638](https://user-images.githubusercontent.com/84202439/118332979-886d6500-b4d0-11eb-9b87-4832eaf6a3b8.jpeg)
![09972660-6D0A-451C-9154-F3FDB37AFD58](https://user-images.githubusercontent.com/84202439/118332984-899e9200-b4d0-11eb-8dff-4034db7e9e7c.jpeg)
also I changed def get_paired_switches(self):
                               switches [ ] to 
                          
                          def get_paired_switches(self):
                                 return [ ]

**Comment by Poohl** [comment]
Ok, that were the pictures I needed, thanks. The first two show some statisctics about the service I presume (I don't have them, maybe my dfeet is outdated). The last picture however is interesting.
The default bluez interface is there (`/org/bluez`) and it's offering all the default interfaces, but there is no controller (`/org/bluez/hciX`). That would suggest you don't have bluetooth, but you stated you do (do the Bluetooth settings in ubuntu work?). 
Does `bluetoothctl show` show someting?

**Comment by JakkFish** [comment]
My Bluetooth settings work I replugged my Bluetooth dongle and I see some other stuff in d-feet
![9FEA57DF-FFE4-42CF-9CB8-8889113AC153](https://user-images.githubusercontent.com/84202439/118334348-03377f80-b4d3-11eb-9031-1083c6bfb2cf.jpeg)
![87960B34-639B-40C0-8972-6D1AA79C154A](https://user-images.githubusercontent.com/84202439/118334355-06cb0680-b4d3-11eb-860f-6abbcef43e5a.jpeg)

Also I used Bluetoothctl in terminal and I see a really long code and it’s still going do you need a picture of that?

oh and also I replaced the code should I try to run the script again?

**Comment by Poohl** [comment]
yep, now the controller shows up. Try it (maybe even without the modification)

**Comment by JakkFish** [comment]
Ok so I tried it and I got this
![30631A43-94A4-412E-AB13-ECDCE3C9224B](https://user-images.githubusercontent.com/84202439/118335249-d97f5800-b4d4-11eb-8fb3-df76701cbc3f.jpeg)

then I removed the = and got this 
![10C7147D-89EE-4E5F-84E7-EA3C933E5E15](https://user-images.githubusercontent.com/84202439/118335296-f4ea6300-b4d4-11eb-8308-101d92cf8372.jpeg)
![C3B3675F-3862-422E-A332-21D5323F780F](https://user-images.githubusercontent.com/84202439/118335309-fae04400-b4d4-11eb-8e20-625973922da5.jpeg)

without modifications it’s same as the first picture I posted

**Comment by Poohl** [comment]
see #4 "how it should be"

**Comment by JakkFish** [comment]
i tried running ExecStart=/usr/lib/bluetooth/bluetoothd -C -P sap,input,avrcp but it says "-C: command not found"

**Comment by JakkFish** [comment]
/lib/systemd/system/bluetooth.service also permission denied and idk what to do from there

**Comment by Poohl** [comment]
Are you root?
(Almost) nothing in this repo works as regular user. Use `sudo su` (not recommended) to permanently become root or prefix a command with `sudo` to run this command as root. To edit a file e.g. `sudo nano /lib/systemd/system/bluetooth.service`.

**Comment by JakkFish** [comment]
I don’t think I’m root I’m using virtual box but did I do this right?

**Comment by JakkFish** [comment]
![0102DAE3-1FA2-48B2-8A76-5995A94076F4](https://user-images.githubusercontent.com/84202439/118343425-90d49880-b4ee-11eb-8c6f-3b57008abeae.jpeg)

**Comment by JakkFish** [comment]
Good news I see this now using return [] or switches = [ ]
![E4574909-E57C-4AA1-95A8-28C9FBBB86C1](https://user-images.githubusercontent.com/84202439/118343494-0ccee080-b4ef-11eb-9b6f-392157d133fd.jpeg)
 Which should I use though? The modify or regular? and sorry for commenting so much thank you for helping me!

**Comment by JakkFish** [comment]
And also if you can answer one last question- How do i change the name? in /etc to Pro Controller

**Comment by Poohl** [comment]
No need to change name. Its done automatically
And you forgot to reload and restart the service.

**Comment by JakkFish** [comment]
The only problem is I have to restart ubuntu every time i run joycontrol because after i stop it or something errors i get this again ![755C24A1-3F49-4704-A81A-E8F957429F01](https://user-images.githubusercontent.com/84202439/118316134-0e32e580-b4bc-11eb-9cf0-ec06a168d676.jpeg)

**Comment by JakkFish** [comment]
other then that everything is working nice!

**Comment by Poohl** [comment]
Ok, after some testing of my own i figured this is some issue with the BlueZ Dbus-API.
Sometimes after restarting the daemon this occurs, and the effect is consintent across Applications. (E.g. d-Feet also just dies when trying to access the Name field)
I'll try to come up with some workaround, but for now this only ever happened on ubuntu and even there is hard for me to reproduce. The main problem with fixing this is it's out of joycontrol's control. And ignoring might not be an option since when the dbus-api is broken there are a few more problems down the road.
For more technical readers: The `Name` property that is expected inside the `Device1` is just absent for some reason and `GetAll` or similar dictionary-lookups just forgo reporting it.
